pilot/pkg/proxy/envoy/v2/util.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v2
+
+import (
+	"sort"
+	"time"
+	// TODO(mostrowski): remove JSON encoding once mixer filter proto spec is available.
+	oldjson "encoding/json"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	http_conn "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/http_connection_manager/v2"
+	google_protobuf "github.com/gogo/protobuf/types"
+	"github.com/golang/protobuf/ptypes/duration"
+)
+
+// normalizeListeners sorts and de-duplicates listeners by address
+func normalizeListeners(listeners []*xdsapi.Listener) []*xdsapi.Listener {
+	out := make([]*xdsapi.Listener, 0, len(listeners))
+	set := make(map[string]bool)
+	for _, listener := range listeners {
+		if !set[listener.Address.String()] {
+			set[listener.Address.String()] = true
+			out = append(out, listener)
+		}
+	}
+	sort.Slice(out, func(i, j int) bool { return out[i].Address.String() < out[j].Address.String() })
+	return out
+}
+
+// mustMarshalToString marshals i to a JSON string. It panics if i cannot be marshaled.
+// TODO(mostrowski): this should be removed once v2 Mixer proto is finalized.
+func mustMarshalToString(i interface{}) string {
+	s, err := oldjson.Marshal(i)
+	if err != nil {
+		panic(err)
+	}
+	return string(s)
+}
+
+// BuildAddress returns a SocketAddress with the given ip and port.
+func BuildAddress(ip string, port uint32) core.Address {
+	return core.Address{
+		Address: &core.Address_SocketAddress{
+			SocketAddress: &core.SocketAddress{
+				Address: ip,
+				PortSpecifier: &core.SocketAddress_PortValue{
+					PortValue: port,
+				},
+			},
+		},
+	}
+}
+
+// GetByAddress returns a listener by its address
+func GetByAddress(listeners []*xdsapi.Listener, addr string) *xdsapi.Listener {
```

Comment:
Where is it used ? Add a TODO to use a map, iterating is not good.

Reply:
It's used in a few places when building protos. I assume you mean to pass the map around (building it inside we still have to iterate). So that would complicate the code a little since we'd have to sync the map with the slice whenever that changes to avoid rebuilding every time. I've put a comment through.